### PR TITLE
Split runtime layout into two configurations: `exec` and `linux`

### PR DESCRIPTION
**Pull request by gaojiaqi7**
- Fix: https://github.com/confidential-containers/td-shim/issues/461

### CODE REVIEW

**Comment by gaojiaqi7** [comment]
This PR depends on PR https://github.com/confidential-containers/td-shim/pull/519, PR https://github.com/confidential-containers/td-shim/pull/516 and PR https://github.com/confidential-containers/td-shim/pull/517

**Comment by jyao1** [comment]
please resolve conflict.

File: td-layout/src/lib.rs [hunk 1]
```diff
@@ -42,6 +42,53 @@ pub const MIN_MEMORY_SIZE: u64 =
 
 pub const TD_PAYLOAD_PARTIAL_ACCEPT_MEMORY_SIZE: u32 = 0x10000000;
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(u32)]
+pub enum E820Type {
```
Comment (jyao1) [inline td-layout/src/lib.rs:34]: why E820 type is related layout ???

File: td-layout/src/lib.rs [hunk 2]
```diff
@@ -29,18 +29,16 @@ use runtime::*;
 
 // Minimal memory size to build the runtime layout.
 #[cfg(feature = "boot-kernel")]
-pub const MIN_MEMORY_SIZE: u64 = (TD_PAYLOAD_ACPI_SIZE
-    + TD_PAYLOAD_UNACCEPTED_MEMORY_BITMAP_SIZE
-    + TD_PAYLOAD_PAGE_TABLE_SIZE
-    + TD_PAYLOAD_EVENT_LOG_SIZE
-    + TD_PAYLOAD_MAILBOX_SIZE) as u64;
+pub const MIN_MEMORY_SIZE: u64 = (ACPI_SIZE
```
Comment (jyao1) [inline td-layout/src/lib.rs]: why rename???

File: td-shim-tools/Cargo.toml [hunk 3]
```diff
@@ -62,7 +62,7 @@ parse_int = { version = "0.6.0", optional = true }
 
 [features]
 default = ["boot-kernel", "enroller", "linker", "signer", "loader", "tee"]
-boot-kernel = ["td-layout/boot-kernel", "td-shim/boot-kernel"]
+boot-kernel = []
```
Comment (jyao1) [inline td-shim-tools/Cargo.toml:65]: why we still keep it?

File: td-layout/src/lib.rs (hunk 1 above)
Comment (gaojiaqi7) [reply td-layout/src/lib.rs:34]: then we can generate E820 table at runtime based on layout config automatically

File: td-layout/src/lib.rs (hunk 2 above)
Comment (gaojiaqi7) [reply td-layout/src/lib.rs]: To align with the name defined in https://github.com/confidential-containers/td-shim/blob/main/td-layout/src/memslice.rs#L16

File: td-shim-tools/Cargo.toml (hunk 3 above)
Comment (gaojiaqi7) [reply td-shim-tools/Cargo.toml:65]: `td-shim-tools` still keeps the feature, will remove it in another PR and use input parameter instead

File: td-shim/src/bin/td-shim/main.rs [hunk 4]
```diff
@@ -182,11 +182,18 @@ fn boot_linux_kernel(
     log_payload_binary(payload, event_log);
     log_payload_parameter(payload_parameter, event_log);
 
+    let mailbox = unsafe {
+        get_dynamic_mem_slice_mut(
```
[the remark this answers is not in the document]
Comment (gaojiaqi7) [reply td-shim/src/bin/td-shim/main.rs]: Yes, the purpose is to remove it in `switch_idt()` and `relocation_mailbox()` functions which are called by `boot_linux_kernel` and `boot_builtin_payload`. Then we don't need to pass the `Memory` instance to it in the changes of following commits

**Comment by gaojiaqi7** [comment]
The patch which removes the `boot-kernel` feature is dropped. And the `E820Type` is kept in the `td-shim` crate

File: td-layout/src/memslice.rs [hunk 5]
```diff
@@ -22,14 +16,14 @@ pub enum SliceType {
...
-    Kernel,
-    /// The `Kernel Parameter` region in runtime memory layout
-    KernelParameter,
     /// The `PAYLOAD` region in runtime memory layout
     Payload,
+    /// The `Kernel Parameter` region in runtime memory layout
+    PayloadParameter,
     /// The `TD_HOB` region in runtime memory layout
     PayloadHob,
+    /// The 'Payload Page Table' region in runtime memory layout
```
Comment (jyao1) [inline td-layout/src/memslice.rs:25]: why we need this?

**Review by xiaoyuxlu — APPROVED** [review]
Looks great.

**Review by xiaoyuxlu — CHANGES_REQUESTED** [review]
Hi Jiaqi, could you update `devtools/td-layout-config/README.md`?

**Comment by gaojiaqi7** [comment]
Updated

**Review by xiaoyuxlu — APPROVED** [review]

**Pull request merged**